tuned/daemon/controller.py

```diff
@@ -373,3 +373,48 @@ def instance_acquire_devices(self, devices, instance_name, caller = None):
 			log.info(rets)
 			return (False, rets)
 		return (True, "OK")
+
+	@exports.export("s", "(bsa(ss))")
+	def get_instances(self, plugin_name, caller = None):
+		"""Return a list of active instances of a plugin or all active instances
+
+		Parameters:
+		plugin_name -- name of the plugin or an empty string
+
+		Return:
+		bool -- True on success
+		string -- error message or "OK"
+		list of string pairs -- [(instance_name, plugin_name)]
+		"""
+		if caller == "":
+			return (False, "Unauthorized", [])
+		active_instances = filter(lambda instance: instance.active, self._daemon._unit_manager.instances)
+		all_pairs = map(lambda instance: (instance.name, instance.plugin.name), active_instances)
```

Comment:
Maybe better to filter first according to the plugin name and then construct the final map? Maybe do all filtering in the first filter lambda together with the active instance. Nevertheless even in this implementation the memory footprint is probably negligible and memory is consumed only during the API call.

Reply:
Makes sense. I refactored it a bit more to make the code prettier.